### PR TITLE
Serde implementations on generated types for Rust

### PR DESCRIPTION
**Pull request by 2kai2kai2**
Added serde implementations to generated types for the Rust SDK, based on the JSON layout used in other SDKs:
- Classes are laid out normal objects, with dynamic fields at the same level (using `#[serde(flatten)]`
- Enums are represented as strings, including dynamic variants (using `#[serde(untagged)]` on the dynamic variant)
- Unions are represented un-tagged (using `#[serde(untagged)]` on the rust enum). This means that serialization+deserialization may be lossy, as the implementation will select the first valid variant for deserialization.

Some types cause errors if you try to deserialize:
- Streaming types cannot be deserialized, similar to how they do not have `BamlEncode` implemented.
- `Types` and `StreamTypes`, the two top-level enumerations of all generated types, cannot be deserialized due to too much ambiguity as to which variant to make: it will almost always be incorrect.
- `DynamicClass`, `DynamicEnum`, and `DynamicUnion` which are the fallback "fully dynamic" versions of these structures. Due to not knowing the associated `name`, we cannot deserialize to these types. Since these only seem to be used in `BamlValue`, and (as noted later) we never deserialize `BamlValue` to a variant containing any of these types, this should not be an issue unless a user directly tries to use one of these types.
- If you try to _directly_ deserialize media types they will error (since the `Deserialize` implementations in the `baml` crate cannot access the runtime directly). However, any time they appear as a field or union variant, we override the field/variant's deserializer to correctly deserialize the media (note: media deserialization could cause funny behavior if we are deserializing a union with multiple media types, where an image could be deserialized as an audio file, for example, due to lack of union disambiguation. Directly deserializing the types would not fix this).

While `BamlValue` does have `Deserialize` implemented, it will map all types into scalars, arrays, and maps. This means that dynamic types will often be typed incorrectly. However, since we do not have any type information on the dynamic types during deserialization, this is the best we can do here.

## Summary by CodeRabbit

* **New Features**
  * Generated Rust types (classes, enums, unions, value containers, streams) now support broad JSON serialization and many deserialization paths.
  * Media objects (image, audio, PDF, video) accept URL or Base64 payloads and deserialize via runtime-backed handlers.
  * Dynamic fields are serialized/deserialized inline (flattened); streaming types gain serialization with explicit guards for ambiguous deserialization.

* **Tests**
  * Added extensive integration tests covering serde (de)serialization and edge cases.

### CODE REVIEW

File: engine/generators/languages/rust/src/generated_types.rs [hunk 1]
```diff
@@ -117,16 +118,22 @@ pub struct FieldRustRendered {
...
     pub type_str: String,
+    pub media_type: Option<MediaTypeRust>,
 }
 
 impl FieldRustRendered {
     pub fn from_field(field: &FieldRust, pkg: &CurrentRenderPackage) -> Self {
+        let media_type = match field.r#type {
+            TypeRust::Media(media_type) => Some(media_type),
+            _ => None,
+        };
```
[the remark this answers is not in the document]
Comment (2kai2kai2) [reply engine/generators/languages/rust/src/generated_types.rs:129]: `media_type: MediaTypeRust` is `Copy`. Since that is the only match arm that binds, this is not an error.

File: languages/rust/baml-macros/src/baml_serde.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    let mut literal_serde = Vec::new();
+
+    for variant in &data.variants {
+        let variant_attrs = VariantAttrs::from_attrs(&variant.attrs)?;
+        let variant_ident = &variant.ident;
+
+        if let Some(ref literal_string) = variant_attrs.literal_string {
+            literal_serde.push(quote! {
+                pub mod #variant_ident {
+                    use ::baml::__internal::serde;
```
[the remark this answers is not in the document]
Comment (aaronvg) [reply languages/rust/baml-macros/src/baml_serde.rs:42]: valid comment -- people's linters may complain
